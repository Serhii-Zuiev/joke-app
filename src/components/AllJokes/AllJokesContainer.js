import React from "react";
import {
  StyledJokesContainer,
  StyledJokesHeading,
  StyledForm,
} from "../../styled";
import { JokesList } from "./";
import { Form, Button } from "react-bootstrap";

const AllJoke = () => {
  return (
    <StyledJokesContainer>
      <StyledJokesHeading>
        Here are our best jokes, but it isn't certain...
      </StyledJokesHeading>

      <Form style={{ width: "6rem" }}>
        <Form.Group controlId="exampleForm.SelectCustom">
          <Form.Label style={{ width: "8rem" }}>choose category</Form.Label>
          <Form.Control as="select" custom>
            <option>All</option>
            <option>Nerdy</option>
            <option>Explicit</option>
          </Form.Control>
        </Form.Group>
      </Form>

      <JokesList />

      <StyledForm>
        <Form.Control
          type="number"
          min="1"
          max="20"
          placeholder="1-20"
          style={{ width: "5rem" }}
        />
        <Button variant="primary" type="submit">
          Get more
        </Button>
      </StyledForm>
    </StyledJokesContainer>
  );
};

export default AllJoke;
