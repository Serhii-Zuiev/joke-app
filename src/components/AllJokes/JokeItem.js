import React from "react";
import { NavLink } from "react-router-dom";
import { Card, Button } from "react-bootstrap";
import { Eyeglasses } from "react-bootstrap-icons";
import { PatchExclamation } from "react-bootstrap-icons";
import { EmojiSmile } from "react-bootstrap-icons";
import { StyledCardContainer } from "../../styled";

function defineJokeCategory(category) {
  let Icon;
  let iconTitle;

  switch (category) {
    case "nerdy":
      Icon = Eyeglasses;
      iconTitle = "Nerdy";
      break;
    case "explicit":
      Icon = PatchExclamation;
      iconTitle = "Explicit";
      break;
    default:
      Icon = EmojiSmile;
      iconTitle = "Funny";

  }
  return { Icon, iconTitle}
}

const JokeItem = ({ text, categories, jokeID }) => {
  
  const category = defineJokeCategory(categories[0])

  return (
    <StyledCardContainer>
      <Card style={{ height: "100%" }}>
        <Card.Body className="cardBody">
          <Card.Text style={{ marginBottom: "0" }}>
            {text.length > 60 ? `${text.slice(0, 60)}...` : text}
          </Card.Text>
          <div className="JokeCategoryWrapper">

            <NavLink  to={{ pathname: `/joke/${jokeID}`, state: { id: jokeID, from: '/' } }}>
              <Button variant="outline-primary">Read joke</Button>
            </NavLink>

            <div className="JokeCategory" title={category.iconTitle}>
              <category.Icon color="#777" size={30} />
            </div>
          </div>
        </Card.Body>
      </Card>
    </StyledCardContainer>
  );
};

export default JokeItem;
