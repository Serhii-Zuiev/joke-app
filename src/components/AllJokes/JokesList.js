import React from "react";
import { StyledJokesList } from "../../styled";
import { JokeItem } from "./";
import { placeholder } from "../../placeholder";

const JokesList = () => {
  return (
    <StyledJokesList>
      {placeholder.map((item) => (
        <li key={item.id} className="jokesListItem">
          <JokeItem text={item.joke} />
        </li>
      ))}
    </StyledJokesList>
  );
};

export default JokesList;
