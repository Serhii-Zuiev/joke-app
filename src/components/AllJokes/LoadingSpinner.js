import React from "react";
import Spinner from "react-bootstrap/Spinner";
import { StyledSpinnerContainer } from "../../styled";

const LoadingSpinner = () => (
  <StyledSpinnerContainer>
    <Spinner animation="border" variant="primary" role="status">
      <span className="sr-only">Loading...</span>
    </Spinner>
  </StyledSpinnerContainer>
);

export default LoadingSpinner;
