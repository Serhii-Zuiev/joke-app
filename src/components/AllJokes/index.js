export { default as AllJokesContainer } from "./AllJokesContainer";
export { default as JokesList } from "./JokesList";
export { default as JokeItem } from "./JokeItem";