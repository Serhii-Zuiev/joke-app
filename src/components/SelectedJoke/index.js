export { default as SelectedJoke } from "./SelectedJoke";