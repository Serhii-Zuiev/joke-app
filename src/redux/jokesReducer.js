export const types = {
  FETCH_JOKES_REQ: "joke-app/reducer/FETCH_JOKES_REQ",
  FETCH_JOKES_OK: "joke-app/reducer/FETCH_JOKES_OK",
  FETCH_JOKES_ERR: "joke-app/reducer/FETCH_JOKES_ERR",
  FETCH_SELECTED_JOKE_REQ: "joke-app/reducer/FETCH_SELECTED_JOKE_REQ",
  START_LOADING: "joke-app/reducer/START_LOADING",
  STOP_LOADING: "joke-app/reducer/STOP_LOADING",
  FILTER_JOKES: "joke-app/reducer/FILTER_JOKES",
};



export const innitialState = {
  jokes: [],
  filter: "All",
  isLoading: false,
  error: null,
};

export default function reducer(state = innitialState, { type, payload }) {
  switch (type) {
    case types.FETCH_JOKES_OK:
      return { ...state, jokes: [...state.jokes, ...payload] };

    case types.FILTER_JOKES:
      return { ...state, filter: payload };

    case types.START_LOADING:
      return { ...state, isLoading: true };

    case types.STOP_LOADING:
      return { ...state, isLoading: false };

    case types.FETCH_JOKES_ERR:
      return { ...state, error: payload };

    default:
      return state;
  }
}



export const fetchJokesRequested = (params) => ({
  type: types.FETCH_JOKES_REQ,
  payload: params,
});

export const fetchSelectedJokeRequested = (params) => ({
  type: types.FETCH_SELECTED_JOKE_REQ,
  payload: params,
});

export const fetchJokesSuccess = (data) => ({
  type: types.FETCH_JOKES_OK,
  payload: data,
});

export const fetchJokesError = (data) => ({
  type: types.FETCH_JOKES_ERR,
  payload: data,
});

export const startLoading = () => ({
  type: types.START_LOADING,
});

export const stopLoading = () => ({
  type: types.STOP_LOADING,
});

export const filterJokes = (filter) => ({
  type: types.FILTER_JOKES,
  payload: filter,
});
