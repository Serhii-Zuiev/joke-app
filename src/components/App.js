import React from "react";
import styled from "styled-components";
import AllJokes from "./AllJokes/AllJokesContainer";
import SelectedJoke from "./SelectedJoke/SelectedJoke";

function App() {
  return (
    <StyledAppContainer>
      <StyledHeader>
        <StyledAppHeading>
          <span role="img" aria-label="face emoji">
            ðŸ˜„
          </span>
          Joke app
        </StyledAppHeading>
      </StyledHeader>

      <AllJokes />
      {/* <SelectedJoke /> */}
    </StyledAppContainer>
  );
}

export default App;

const StyledAppContainer = styled.div`
  background-color: #eee;
`;

const StyledHeader = styled.header`
  height: 70px;
  width: 90%;
  margin: 0 auto;
`;

const StyledAppHeading = styled.h1`
  font-size: 26px;
  color: #333;
  padding: 16px 0;
  letter-spacing: 1px;
  text-align: center;
  border-bottom: 2px solid #777;
`;
