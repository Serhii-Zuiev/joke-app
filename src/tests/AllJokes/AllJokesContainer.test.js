import React from "react";
import { Provider } from "react-redux";
import { createStore } from "redux";

import { AllJokesContainer } from "../../components/AllJokes";
import { default as reducer, innitialState } from "../../redux/jokesReducer";
import { filterJokes, fetchJokesRequested } from "../../redux/jokesReducer";

describe("<AllJokesContainer /> unit test", () => {
    const mockStore = createStore(reducer, innitialState);
    mockStore.dispatch = jest.fn();
    const getWrapper = () =>
        mount(
            <Provider store={mockStore}>
                <AllJokesContainer />
            </Provider>
        );

    const wrapper = getWrapper();

    it("should render AllJokesContainer component", () => {
        expect(wrapper).toMatchSnapshot();
    });

    it("should dispatch initial fetch jokes action", () => {
        getWrapper();
        expect(mockStore.dispatch).toHaveBeenCalledWith(
            fetchJokesRequested(20)
        );
    });

    it("select should dispatch the correct action on change", () => {
        wrapper.find("select").simulate("change");
        expect(mockStore.dispatch).toHaveBeenCalledWith(filterJokes("All"));
    });

    it("form should dispatch the correct action on submit", () => {
        wrapper
            .find(".form-control")
            .simulate("change", { target: { value: 11 } });
        wrapper.find(".sc-Axmtr").simulate("submit");
        expect(mockStore.dispatch).toHaveBeenCalledWith(
            fetchJokesRequested(11)
        );
    });
});
