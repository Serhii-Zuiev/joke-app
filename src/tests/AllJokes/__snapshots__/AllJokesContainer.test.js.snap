// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AllJokesContainer /> unit test should render AllJokesContainer component 1`] = `
<Provider
  store={
    Object {
      "dispatch": [MockFunction],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <AllJoke>
    <styled.section>
      <section
        className="sc-AxgMl gTGhQd"
      >
        <styled.h2>
          <h2
            className="sc-AxheI etBnmE"
          >
            Here are our best jokes, but it isn't certain...
          </h2>
        </styled.h2>
        <Form
          inline={false}
          style={
            Object {
              "width": "6rem",
            }
          }
        >
          <form
            className=""
            style={
              Object {
                "width": "6rem",
              }
            }
          >
            <FormGroup
              controlId="exampleForm.SelectCustom"
            >
              <div
                className="form-group"
              >
                <FormLabel
                  column={false}
                  srOnly={false}
                  style={
                    Object {
                      "width": "8rem",
                    }
                  }
                >
                  <label
                    className="form-label"
                    htmlFor="exampleForm.SelectCustom"
                    style={
                      Object {
                        "width": "8rem",
                      }
                    }
                  >
                    choose category
                  </label>
                </FormLabel>
                <FormControl
                  as="select"
                  custom={true}
                  onChange={[Function]}
                  value="All"
                >
                  <select
                    className="custom-select"
                    id="exampleForm.SelectCustom"
                    onChange={[Function]}
                    value="All"
                  >
                    <option>
                      All
                    </option>
                    <option>
                      Nerdy
                    </option>
                    <option>
                      Explicit
                    </option>
                  </select>
                </FormControl>
              </div>
            </FormGroup>
          </form>
        </Form>
        <JokesList
          allJokes={Array []}
        >
          <styled.ul>
            <ul
              className="sc-AxmLO cZbuqf"
            />
          </styled.ul>
        </JokesList>
        <styled.form
          onSubmit={[Function]}
        >
          <form
            className="sc-Axmtr jRulXR"
            onSubmit={[Function]}
          >
            <FormControl
              max="20"
              min="1"
              onChange={[Function]}
              placeholder="1-20"
              required={true}
              style={
                Object {
                  "width": "5rem",
                }
              }
              type="number"
              value=""
            >
              <input
                className="form-control"
                max="20"
                min="1"
                onChange={[Function]}
                placeholder="1-20"
                required={true}
                style={
                  Object {
                    "width": "5rem",
                  }
                }
                type="number"
                value=""
              />
            </FormControl>
            <Button
              active={false}
              disabled={false}
              type="submit"
              variant="primary"
            >
              <button
                className="btn btn-primary"
                disabled={false}
                type="submit"
              >
                Get more
              </button>
            </Button>
          </form>
        </styled.form>
      </section>
    </styled.section>
  </AllJoke>
</Provider>
`;
